dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java

```diff
@@ -67,32 +71,64 @@ public List<BlockWorkerInfo> getPreferredWorkers(List<BlockWorkerInfo> blockWork
   private static class ConsistentHashProvider {
     private static final HashFunction HASH_FUNCTION = murmur3_32_fixed();
     private static final int MAX_ATTEMPTS = 100;
-    private List<BlockWorkerInfo> mLastWorkerInfos = ImmutableList.of();
-    private NavigableMap<Integer, BlockWorkerInfo> mActiveNodesByConsistentHashing;
-
-    private volatile long mLastUpdatedTimestamp = 0L;
-
-    private final AtomicBoolean mNeedUpdate = new AtomicBoolean(false);
-
-    private static final long WORKER_INFO_UPDATE_INTERVAL_MS = 1000L;
-
+    private static final long WORKER_INFO_UPDATE_INTERVAL_NS = Constants.SECOND_NANO;
+    private volatile List<BlockWorkerInfo> mLastWorkerInfos = ImmutableList.of();
+    // Can only be null before the first call to refresh
+    @Nullable
+    private volatile NavigableMap<Integer, BlockWorkerInfo> mActiveNodesByConsistentHashing;
+    private final Semaphore mInitLock = new Semaphore(1);
+
+    // Must use System.nanoTime to ensure monotonic increment
+    private final AtomicLong mLastUpdatedTimestamp = new AtomicLong(System.nanoTime());
+
+    /**
+     * Initializes or refreshes the worker list using the given list of workers and number of
+     * virtual nodes.
+     */
     public void refresh(List<BlockWorkerInfo> workerInfos, int numVirtualNodes) {
-      // check if we need to update worker info
-      if (mLastUpdatedTimestamp <= 0L
-          || System.currentTimeMillis() - mLastUpdatedTimestamp > WORKER_INFO_UPDATE_INTERVAL_MS) {
-        mNeedUpdate.set(true);
+      Preconditions.checkArgument(!workerInfos.isEmpty(),
+          "cannot refresh hash provider with empty worker list");
+      // When the active nodes map does not exist, the hash provider is not initialized yet.
+      // let one caller initialize the map while blocking all others.
+      if (mActiveNodesByConsistentHashing == null) {
+        mInitLock.acquireUninterruptibly();
+        // only one thread should reach here
+        // test again to skip re-initialization
+        try {
+          if (mActiveNodesByConsistentHashing == null) {
+            mActiveNodesByConsistentHashing = build(workerInfos, numVirtualNodes);
+            mLastWorkerInfos = workerInfos;
+            mLastUpdatedTimestamp.set(System.nanoTime());
+          }
+        } finally {
+          mInitLock.release();
+        }
       }
-      // update worker info if needed
-      if (mNeedUpdate.compareAndSet(true, false)) {
-        if (isWorkerInfoUpdated(workerInfos, mLastWorkerInfos)) {
-          build(workerInfos, numVirtualNodes);
+      // check if the worker list has expired
+      long lastUpdateTs = mLastUpdatedTimestamp.get();
+      long currentTs = System.nanoTime();
```

Comment:
https://stackify.com/heres-how-to-calculate-elapsed-time-in-java/ 
Actually I think `Instant` may be better? You may be getting the nano time from different threads. I don't think that gives you a strictly monotonous behavior?

Reply:
The javadoc of this `System.nanoTime()` is somewhat vague:
```
This method provides namesecond precision, but not necessarily nanosecond resolution (that is, how frequently the value changes).
...
The value returend by this method becomes meaningful only when the difference between two such values, obtained within the same instance of a Java virtual machine, is computed.
```

My interpretation is: from the same JVM, if multiple threads get `System.nanoTime()`, the result will be monotonous but possibly stale (seeing twice for the same timestamp). The monotocity is broken ONLY when the values are from different JVMs.

That said, I guess our use here is okay.

`Instant` uses `System.clockTime()` so can be impacted by changing system time. Also the extra thread safety may not be necessary (and is less performant).